onedal/_device_offload.py

```diff
@@ -194,17 +219,21 @@ def support_usm_ndarray(freefunc=False, queue_param=True):
 
     def decorator(func):
         def wrapper_impl(obj, *args, **kwargs):
-            usm_iface = _extract_usm_iface(*args, **kwargs)
+            usm_iface, input_array_api, input_array_api_device = _extract_array_attr(
+                *args, **kwargs
+            )
             data_queue, hostargs, hostkwargs = _get_host_inputs(*args, **kwargs)
             if queue_param and not (
                 "queue" in hostkwargs and hostkwargs["queue"] is not None
             ):
                 hostkwargs["queue"] = data_queue
             result = _run_on_device(func, obj, *hostargs, **hostkwargs)
-            if usm_iface is not None and hasattr(result, "__array_interface__"):
+            if usm_iface is not None:
                 result = _copy_to_usm(data_queue, result)
                 if dpnp_available and len(args) > 0 and isinstance(args[0], dpnp.ndarray):
                     result = _convert_to_dpnp(result)
+            elif input_array_api and not _is_numpy_namespace(input_array_api):
```

Comment:
Only check for array_api stuff if usm_iface is None.  Otherwise we are doing unnecessary work.

Reply:
Absolutely not, it is actually required to avoid numpy namespace here, and leave the input as is if numpy namespace.

Reply:
The ```elif``` here comes after this call: https://github.com/samir-nasibli/scikit-learn-intelex/blob/enh/functional_array_api/onedal/_device_offload.py#L231(```if usm_iface is not None:```) Thus in this function, when ```usm_iface``` is not None, we have wasted cycles discovering ```input_array_api``` and ```input_array_api_device``` from ```_extract_array_attr```. You should have something from ```_array_api.py``` which returns just that necessary info when the Line 231 if statement evaluates vales (i.e. ```usm_iface is None```)

Reply:
Thank you @icfaust ! It was good point. I have already updated the logic for all handlers and helper functions. Only on `_transfer_to_host` we do check's for both `usm_iface` and `array_api`. Currently this means, that we can pass both usm and array api inputs, but the data queue from usm_data or explicitly `queue` passed will be used for device offloading.